tests/tools/fswalker/fswalker.go

```diff
@@ -0,0 +1,200 @@
+func NewWalkCompare() *WalkCompare {
+	return &WalkCompare{
+		GlobalFilterMatchers: []string{
+			"ctime:",
+			"atime:",
+			"mtime:",
+		},
+	}
+}
+
+// Gather meets the checker.Comparer interface. It performs a fswalker Walk
+// and returns the resulting Walk as a protobuf Marshaled buffer.
+func (chk *WalkCompare) Gather(ctx context.Context, path string) ([]byte, error) {
+	walkData, err := walker.WalkPathHash(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+
+	err = rerootWalkDataPaths(walkData, path)
+	if err != nil {
+		return nil, err
+	}
+
+	// Store the walk data along with the snapshot ID
+	b, err := proto.Marshal(walkData)
+	if err != nil {
+		return nil, err
+	}
+
+	return b, nil
+}
+
+// Compare meets the checker.Comparer interface. It performs a fswalker Walk
+// on the provided file path, unmarshals the comparison data as a fswalker Walk,
+// and generates a fswalker report comparing the two Walks. If there are any differences
+// an error is returned, and the full report will be written to the provided writer
+// as JSON.
+func (chk *WalkCompare) Compare(ctx context.Context, path string, data []byte, reportOut io.Writer) error {
+	beforeWalk := &fspb.Walk{}
+
+	err := proto.Unmarshal(data, beforeWalk)
+	if err != nil {
+		return err
+	}
+
+	afterWalk, err := walker.WalkPathHash(ctx, path)
+	if err != nil {
+		return err
+	}
+
+	err = rerootWalkDataPaths(afterWalk, path)
+	if err != nil {
+		return err
+	}
+
+	report, err := reporter.Report(ctx, &fspb.ReportConfig{}, beforeWalk, afterWalk)
+	if err != nil {
+		return err
+	}
+
+	chk.filterReportDiffs(report)
+
+	err = validateReport(report)
+	if err != nil {
+		printReportSummary(report, reportOut)
+
+		b, marshalErr := json.MarshalIndent(report, "", "   ")
+		if marshalErr != nil {
+			_, reportErr := reportOut.Write([]byte(marshalErr.Error()))
+			if reportErr != nil {
+				return fmt.Errorf("error while writing marshal error (%v): %v", marshalErr.Error(), reportErr.Error())
+			}
+
+			return marshalErr
+		}
+
+		if _, wrErr := reportOut.Write(b); wrErr != nil {
+			return err
+		}
+
+		return err
+	}
+
+	return nil
+}
+
+func printReportSummary(report *fswalker.Report, reportOut io.Writer) {
+	rptr := &fswalker.Reporter{}
+	rptr.PrintDiffSummary(reportOut, report)
+	rptr.PrintReportSummary(reportOut, report)
+	rptr.PrintRuleSummary(reportOut, report)
+}
+
+func (chk *WalkCompare) filterReportDiffs(report *fswalker.Report) {
+	var newModList []fswalker.ActionData
+
+	for _, mod := range report.Modified {
+		var newDiffItemList []string
+
+		diffItems := strings.Split(mod.Diff, "\n")
+
+	DiffItemLoop:
+		for _, diffItem := range diffItems {
+			for _, filterStr := range chk.GlobalFilterMatchers {
+				if strings.Contains(diffItem, filterStr) {
+					fmt.Printf("FILTERING %s due to filtered prefix %q\n", diffItem, filterStr)
+					continue DiffItemLoop
+				}
+			}
+
+			// Filter the rename of the root directory
+			if isRootDirectoryRename(diffItem, mod) {
```

Comment:
if the walk was re-rooted, does `isRootDirectoryRename()` return true?

Reply:
`isRootDirectoryRename` just returns true if the reporter finds a difference between two walks that is the topmost directory on the walk was renamed. For instance if I `kopia snapshot create /some_data_dir`, I do the walk starting from and including `some_data_dir`. Then when I do `kopia snapshot restore <snapID> /some_target_dir`, I do the walk starting from `some_target_dir`. I expect everything to be the same, except that the name of that directory is whatever I restored to. Does that answer your question?

Reply:
Partly. So the report will only contain a single difference in that case. Or will all the files be flagged as different because of the rename?

Reply:
The report contains a list of file system entries that were modified between the before walk and after walk. For each modified file/dir, there is a list of changes that the reporter found. The root directory could be expected to contain one difference: that directory's name may have changed. It might also have undergone other differences, permissions changed etc. `isRootDirectoryRename` will cause only the rename difference to be filtered. If there are any other differences on this entry, for instance if the permissions changed, they will be flagged as a failure.

All other child files/dirs have no dependency on the rename, because all of their paths are relative to the source directory (due to `rerootWalkDataPaths`), and thus their names, paths, everything, are expected to match.

So the report might say there's a modification due to the root directory at path `"."` had its "name" field changed from `source-dir` to `target-dir`, but the `file.txt` at path `"./file.txt"` should still have the same fields (name, path, size, owner, permissions, etc.) before the snapshot and after the restore. `isRootDirectoryRename` is intended to filter out only the root directory rename, since it is expected, and nothing more.